nixos/modules/services/home-automation/esphome.nix

```diff
@@ -0,0 +1,153 @@
+    };
+
+    openFirewall = mkOption {
+      default = false;
+      type = types.bool;
+      description = mdDoc "Whether to open the firewall for the specified port.";
+    };
+
+    allowedDevices = mkOption {
+      default = [];
+      example = [{ node = "/dev/serial/by-id/usb-Silicon_Labs_CP2102_USB_to_UART_Bridge_Controller_0001-if00-port0"; modifier = "rw"; }];
+      description = lib.mdDoc ''
+        A list of device nodes to which {command}`esphome` has access to.
+        Beware that permissions are not added dynamically when a device
+        is plugged in while the service is already running.
+      '';
+      type = types.listOf (types.submodule {
+        options = {
+          node = mkOption {
+            example = "/dev/ttyUSB*";
+            type = types.str;
+            description = lib.mdDoc "Path to device node";
+          };
+          modifier = mkOption {
+            example = "rw";
+            type = types.str;
+            description = lib.mdDoc ''
+              Device node access modifier. Takes a combination
+              `r` (read), `w` (write), and `m` (mknod). See the
+              `systemd.resource-control(5)` man page for more
+              information.
+            '';
+          };
+        };
+      });
+    };
+  };
+
+  config = mkIf cfg.enable {
+    networking.firewall.allowedTCPPorts = mkIf (cfg.openFirewall && !cfg.enableUnixSocket) [cfg.port];
+
+    systemd.services.esphome = {
+      description = "ESPHome dashboard";
+      after = ["network.target"];
+      wantedBy = ["multi-user.target"];
+      path = [cfg.package];
+
+      # platformio fails to determine the home directory when using DynamicUser
+      environment.PLATFORMIO_CORE_DIR = "${stateDir}/.platformio";
+
+      serviceConfig = {
+        ExecStart = "${cfg.package}/bin/esphome dashboard ${esphomeParams} ${stateDir}";
+        DynamicUser = true;
+        User = "esphome";
+        Group = "esphome";
+        WorkingDirectory = stateDir;
+        StateDirectory = "esphome";
+        StateDirectoryMode = "0750";
+        Restart = "on-failure";
+        RuntimeDirectory = mkIf cfg.enableUnixSocket "esphome";
+        RuntimeDirectoryMode = "0750";
+
+        # Hardening
+        CapabilityBoundingSet = "";
+        LockPersonality = true;
+        MemoryDenyWriteExecute = true;
+        DevicePolicy = "closed";
+        DeviceAllow = map (d: "${d.node} ${d.modifier}") cfg.allowedDevices;
+        SupplementaryGroups = ["dialout"];
+        NoNewPrivileges = true;
+        PrivateUsers = true;
+        PrivateTmp = true;
+        ProtectClock = true;
+        ProtectControlGroups = true;
+        ProtectHome = true;
+        ProtectHostname = true;
+        ProtectKernelLogs = true;
+        ProtectKernelModules = true;
+        ProtectKernelTunables = true;
+        ProtectProc = "invisible";
+        ProcSubset = "pid";
+        ProtectSystem = "strict";
+        RemoveIPC = true;
```

Comment:
From the systemd-exec man page https://www.freedesktop.org/software/systemd/man/systemd.exec.html#DynamicUser= : 

> If DynamicUser= is enabled, RemoveIPC= and PrivateTmp= are implied (and cannot be turned off). 

I think we can remove these two declarations here.

Reply:
Also, from that same page:
> Furthermore NoNewPrivileges= and RestrictSUIDSGID= are implicitly enabled (and cannot be disabled), 
So that could be removed as well.

Reply:
I included them to indicate that they were thought of. Should I maybe just comment them out and add a hint that they are implied?

Reply:
Done :)